File: leetcode46.cpp
#include <vector>
using namespace std;

class Solution {
public:
	vector<vector<int>> res;
    void dfs(int index,int n,vector<int>& arr,vector<bool>& hash,vector<int>& nums)
	{	
		if(index == n){
			res.push_back(arr);
			return ;
		}
		for(int i = 0;i <  n;++i)
		{
			if(hash[i]) continue;
			hash[i] = true;
			arr.push_back(nums[i]);
			dfs(index + 1,n,arr,hash,nums);
			arr.pop_back();
			hash[i] = false;
		}


	}	

	vector<vector<int>> permute(vector<int>& nums) {
		int n = nums.size();
		vector<bool> hash(n,false);
		vector<int> arr;
		dfs(0,n,arr,hash,nums);
		return  res;
    }
};

File: leetcode46.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */

var permuteUnique = function(nums) {
    const n = nums.length;
    if(n === 1) return [nums];
    nums.sort((a,b)=> a - b);
    const res = [];
    const temp = [];
    const pass = new Array(n).fill(false); // 判断是否访问过
    function dfs(current){
        if(current == n){
            res.push([...temp]);
            return;
        }
        for(let i = 0;i < n;++i){
            if(pass[i] || i && nums[i] == nums[i-1] && !pass[i-1]) continue;
            temp[current] = nums[i];
            pass[i] = true;
            dfs(current + 1);
            pass[i] = false;
        }
    }
    dfs(0);
    return res;

};